wp-includes/class-wp-service-workers.php

```diff
@@ -83,6 +220,170 @@ public function register( $handle, $src, $deps = array(), $scope = self::SCOPE_A
 		return parent::add( $handle, $src, $deps, false, compact( 'scope' ) );
 	}
 
+	/**
+	 * Register route and caching strategy.
+	 *
+	 * @todo Figure out a better name for $route since it can also be a list.
+	 * @param string|array $route Route or list of routes.
+	 * @param string       $strategy Strategy, can be WP_Service_Workers::STRATEGY_STALE_WHILE_REVALIDATE, WP_Service_Workers::STRATEGY_CACHE_FIRST,
+	 *                         WP_Service_Workers::STRATEGY_NETWORK_FIRST, WP_Service_Workers::STRATEGY_CACHE_ONLY,
+	 *                         WP_Service_Workers::STRATEGY_NETWORK_ONLY, WP_Service_Workers::::STRATEGY_PRECACHE.
+	 * @param array        $args Array of args, can include cache_name, max_age, max_entries.
+	 */
+	public function register_cached_route( $route, $strategy, $args = array() ) {
+
+		if ( ! in_array( $strategy, array(
+			self::STRATEGY_STALE_WHILE_REVALIDATE,
+			self::STRATEGY_CACHE_FIRST,
+			self::STRATEGY_CACHE_ONLY,
+			self::STRATEGY_NETWORK_FIRST,
+			self::STRATEGY_NETWORK_ONLY,
+			self::STRATEGY_PRECACHE,
+		), true ) ) {
+			_doing_it_wrong( __METHOD__, esc_html__( 'Strategy must be either WP_Service_Workers::STRATEGY_STALE_WHILE_REVALIDATE, WP_Service_Workers::STRATEGY_CACHE_FIRST, WP_Service_Workers::STRATEGY_NETWORK_FIRST, WP_Service_Workers::STRATEGY_CACHE_ONLY, or WP_Service_Workers::STRATEGY_NETWORK_ONLY.', 'pwa' ), '0.2' );
+			return;
+		}
+
+		if ( self::STRATEGY_PRECACHE === $strategy ) {
+			if ( ! is_array( $route ) ) {
+				$route = array( $route );
+			}
+
+			$this->registered_caching_routes[] = array(
+				'route'    => $route, // @todo route is Not the best name in case of array of routes for precaching.
+				'strategy' => $strategy,
+			);
+		} elseif ( ! is_string( $route ) ) {
+			/* translators: %s is caching strategy */
+			$error = sprintf( __( 'Route for the caching strategy %s must be a string.', 'pwa' ), $strategy );
+			_doing_it_wrong( __METHOD__, esc_html( $error, 'pwa' ), '0.2' );
+		} else {
+
+			// Set default to string literal for route.
+			if ( ! isset( $args['regex'] ) ) {
+				$args['regex'] = false;
+			}
+
+			$this->registered_caching_routes[] = array(
+				'route'    => $route,
+				'strategy' => $strategy,
+				'args'     => $args,
+			);
+		}
+	}
+
+	/**
+	 * Register precaching for routes. Gets the hashes of files' contents and adds as revision for each route.
+	 *
+	 * @param array $routes Array of routes.
+	 * @return string Precaching logic.
+	 */
+	protected function register_precaching_for_routes( $routes ) {
+
+		$routes_list = '[';
```

Comment:
It would be better to construct an array here, like `$routes_list = array()`. And then for each `$route` do:

```php
$routes_list[] = array(
    'url' => $base_url . $route,
    'revision' => $revision,
);
```

And then at the end just do:

```php
return sprintf( "wp.serviceWorker.precaching.precacheAndRoute( %s );\n", wp_json_encode( $routes_list ) );
```